lib/PromAlertProxy/Hub.pm

```diff
@@ -0,0 +1,186 @@
+  state $check = compile(Object, Str, Int); # XXX IPAddress, Port
+  $check->(@_);
+
+  my $http = Net::Async::HTTP::Server::PSGI->new(app => $self->_app);
+  $self->loop->add($http);
+
+  $http->listen(
+    addr => {
+      family   => 'inet',
+      socktype => 'stream',
+      ip       => $ip,
+      port     => $port,
+    },
+    on_listen_error => sub ($err, @) {
+      $Logger->log_fatal(["couldn't listen on %s:%s: %s", $ip, $port, $err]);
+    },
+  );
+
+  $Logger->log(["listening on %s:%s", $ip, $port]);
+}
+
+has _targets        => ( is => 'rw', default => sub { {} } );
+has _default_target => ( is => 'rw' );
+
+sub add_target ($self, $target) {
+  state $check = compile(Object, role_type('PromAlertProxy::Target'));
+  $check->(@_);
+
+  $self->_targets->{$target->id} = $target;
+  $Logger->log(['added target: %s', $target->id]);
+
+  if ($target->default) {
+    if ($self->_default_target) {
+      $Logger->log([
+        "can't make '%s' the default target, '%s' is already the default",
+        $target->id, $self->_default_target->id,
+      ]);
+    }
+    else {
+      $self->_default_target($target);
+      $Logger->log(['setting default target: %s', $target->id]);
+    }
+  }
+}
+
+sub dispatch ($self, $alert) {
+  state $check = compile(Object, class_type('PromAlertProxy::PromAlert'));
+  $check->(@_);
+
+  $Logger->log(["received alert: %s (%s)", $alert->name, $alert->summary]);
+
+  $self->_metrics->inc('promalertproxy_alerts_received_total', { name => $alert->name });
+
+  my $use_default = 0;
+  my $target_name = $alert->target;
+  if (!$target_name || !$self->_targets->{$target_name}) {
+    $use_default = 1;
+  }
+
+  my $target = $use_default ? $self->_default_target : $self->_targets->{$target_name};
+  if ($target) {
+    $Logger->log(["%s: dispatching to target '%s'%s", $alert->name, $target->id, $use_default ? " (as default)" : ""]);
+    $self->_raise_safe($alert, $target);
+    return;
+  }
+
+  $Logger->log(["%s: no target or default found, fallback dispatching to all", $alert->name]);
+  $self->_raise_all($alert);
+}
+
+sub _raise_safe ($self, $alert, $target) {
+  $self->_metrics->inc('promalertproxy_alerts_dispatched_total', { name => $alert->name, target => $target->id });
+
+  my $failed = eval {
+    $target->raise($alert);
+  };
+
+  if (my $err = $@) {
+    $Logger->log(["%s: dispatch to target '%s' crashed: %s", $alert->name, $target->id, $err]);
+    $failed = 1;
+  }
+
+  if ($failed) {
+    $Logger->log(["%s: dispatch to target '%s' failed, fallback dispatching to all", $alert->name, $target->id]);
+    $self->_raise_all($alert);
+  }
+}
+
+sub _raise_all ($self, $alert) {
+  $self->_metrics->inc('promalertproxy_alerts_fallback_dispatched_total', { name => $alert->name });
+
+  for my $target (values $self->_targets->%*) {
+    $Logger->log(["%s: fallback dispatching to target '%s'", $alert->name, $target->id]);
+
+    my $failed = eval {
+      $target->raise($alert);
```

Comment:
Traditionally its so you can write:

```c
if (!do_something()) {
    printf("failed to do something\n");
}
```

but for sure.

tbh, I could just die, since its always in eval, but I wanted that in case the subthing crashes out.

What I really want is `Result<T>`. Sigh.